src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -2515,6 +2563,98 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
             return true;
         }
 
+        private int FindLineSeparatorMultiSegment(ReadOnlySpan<byte> localBuffer, ref int dangerousLineSeparatorBytesConsumed)
+        {
+            Debug.Assert(dangerousLineSeparatorBytesConsumed >= 0 && dangerousLineSeparatorBytesConsumed <= 2);
+
+            if (dangerousLineSeparatorBytesConsumed != 0)
+            {
+                ThrowOnDangerousLineSeparatorMultiSegment(localBuffer, ref dangerousLineSeparatorBytesConsumed);
+
+                if (dangerousLineSeparatorBytesConsumed != 0)
+                {
+                    // this can only happen if localBuffer size is 1 and we have previously only consumed 1 byte
+                    // or localBuffer is 0
+                    Debug.Assert(dangerousLineSeparatorBytesConsumed >= 1 && dangerousLineSeparatorBytesConsumed <= 2 && localBuffer.Length <= 1);
+                    return -1;
+                }
+            }
+
+            int totalIdx = 0;
+            while (true)
+            {
+                int idx = localBuffer.IndexOfAny(JsonConstants.LineFeed, JsonConstants.CarriageReturn, JsonConstants.MaybeDangerousLineSeparator);
+                dangerousLineSeparatorBytesConsumed = 0;
+
+                if (idx == -1)
+                    return -1;
+
+                if (localBuffer[idx] != JsonConstants.MaybeDangerousLineSeparator)
+                    return totalIdx + idx;
+
+                int p = idx + 1;
+                localBuffer = localBuffer.Slice(p);
+                totalIdx += p;
+
+                // or dangerousLineSeparatorBytesConsumed = 1
+                dangerousLineSeparatorBytesConsumed++;
+                ThrowOnDangerousLineSeparatorMultiSegment(localBuffer, ref dangerousLineSeparatorBytesConsumed);
+
+                if (dangerousLineSeparatorBytesConsumed != 0)
+                {
+                    // this can only happen in the end of stream
+                    Debug.Assert(localBuffer.Length < 2);
+                    return -1;
+                }
+            }
+        }
+
+        // assumes first byte (JsonConstants.MaybeDangerousLineSeparator) is already read
+        private void ThrowOnDangerousLineSeparatorMultiSegment(ReadOnlySpan<byte> localBuffer, ref int dangerousLineSeparatorBytesConsumed)
```

Comment:
This seems unnecessarily complicated, especially compared to the single-segment case. Why do we have a loop here and why do we need to keep passing the int by ref?

Reply:
we need to keep track of how many bytes we consumed in the last segment and I didn't want to put GetNextSpan as it would make it even more complicated. I have removed the loop which hopefully made it a bit more legible